libkmod/libkmod-module.c

```diff
@@ -509,23 +506,17 @@ KMOD_EXPORT int kmod_module_get_filtered_blacklist(const struct kmod_ctx *ctx,
 	return kmod_module_apply_filter(ctx, KMOD_FILTER_BLACKLIST, input, output);
 }
 
-static const struct kmod_list *module_get_dependencies_noref(const struct kmod_module *mod)
+static void module_get_dependencies_noref(const struct kmod_module *mod)
 {
 	if (!mod->init.dep) {
 		/* lazy init */
 		char *line = kmod_search_moddep(mod->ctx, mod->name);
 
-		if (line == NULL)
-			return NULL;
-
-		kmod_module_parse_depline((struct kmod_module *)mod, line);
-		free(line);
-
-		if (!mod->init.dep)
-			return NULL;
+		if (line != NULL) {
```

Comment:
Don't think we need the NULL check here. Both `strchr(NULL, ':');` and `free(NULL)` are safe.

Reply:
What makes you believe that `strchr(NULL, ':')` is safe? Definitely not the case for glibc and not part of the standard. Just make sure that the compiler does not outsmart you:

```
cat > search.c << EOF
#include <string.h>

char *search(char *p)
{
        return strchr(p, ':');
}
EOF
cat > poc.c << EOF
#include <stddef.h>

extern char *search(char *);

int
main(int argc, char **argv)
{
        search(NULL);
        return 0;
}
EOF
cc -c test.c poc.c
cc -o poc test.o poc.o
./poc
```

leads to

```
Segmentation fault (core dumped)
```

Reply:
I failed at reading `man 3 strchr` it seems. Thanks for the correction :bow: